### PR TITLE
Websocket 2.1.0 release

### PR DESCRIPTION
**Pull request by markt-asf**
## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [X] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Generated IP Log
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [X] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/websocket/jakarta.websocket-api/2.1.0/
      https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/websocket/jakarta.websocket-client-api/2.1.0/
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee10/staged/eftl/jakarta-websocket-tck-2.1.0.zip
- [X] The URL of the compatibility certification request issue:
      https://github.com/jakartaee/websocket/issues/396
- [X] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by kazumura** [comment]
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request

10. If a Release Review is required, the specification project team contacts the EMO to initiate the review by sending an email to emo@eclipse.org.
    (A Release Review is not required if the current release is a Service Release based on a previously successful Major or Minor
    release as indicated by a release record on the project's Releases page, e.g., the [Jakarta Servlet releases page](https://projects.eclipse.org/projects/ee4j.servlet/reviews).)
    - [ ] An issue will be created by the EMO to track the release review.
    - [ ] The specification project team requests approval for the release from the PMC by sending an email to ee4j-pmc@eclipse.org.
    - [ ] The specification project team then delivers an IP Log to the IP Team for their review as described in https://www.eclipse.org/projects/handbook/#pmi-commands-iplog.

11. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Review by kazumura — CHANGES_REQUESTED** [review]
@markt-asf 
The link of apidocs is incorrect.  
If you have two apidocs, you may need two links, one for client the other for server.

**Comment by markt-asf** [comment]
@kazumura
Updated as requested

File: websocket/2.1/apidocs/client/constant-values.html [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+<noscript>
+<div>JavaScript is disabled on your browser.</div>
+</noscript>
+</div>
+<a id="skip.navbar.bottom">
+<!--   -->
+</a></div>
+<!-- ======== END OF BOTTOM NAVBAR ======= -->
+</nav>
+<p class="legalCopy"><small>Comments to: <a href="mailto:websocket-dev@eclipse.org">websocket-dev@eclipse.org</a>.<br> Copyright &#169; 2018, 2020 Eclipse Foundation. All rights reserved.<br> Use is subject to <a href="./doc-files/EFSL.html" target="_top">license terms</a>.</small></p>
```
Comment (kazumura) [inline websocket/2.1/apidocs/client/constant-values.html:179]: The years of copyright are "2018, 2020", but the ones of the specification documents are "2018, 2021". They are inconsistent.

Comment (markt-asf) [reply websocket/2.1/apidocs/client/constant-values.html:179]: I have fixed this but there is an issue with the CI system at the moment preventing the build completing. As soon as that is fixed, I'll update API docs.

**Comment by markt-asf** [comment]
@kazumura Javadoc updated

**Comment by kazumura** [comment]
@markt-asf 
Thanks for updating the TCK.  
I would like ask you some changes for the TCK documents.

* Please update README.txt
  * Use Websocket version 2.1 but not 2.0
  * Use Java SE 11 but not 8

* Please update TCK User's Guide
  * Use Websocket version 2.1 but not 2.0

**Review by kazumura — APPROVED** [review]

**Comment by kazumura** [comment]
@markt-asf 
Thanks for updaring TCK documents.  
The remaining is CCR.

**Comment by kazumura** [comment]
@markt-asf 
Do you have any idea when we can submit CCR ?

**Comment by markt-asf** [comment]
That would be a question for the Glassfish project. I'm not tracking that project's progress.

**Comment by scottmarlow** [comment]
https://github.com/eclipse-ee4j/jakartaee-tck/issues/873 is the challenge

**Comment by markt-asf** [comment]
The challenge is not valid but the WebSocket team members can't close the issue.

**Comment by kazumura** [comment]
@markt-asf @jansupol 
Thanks for submitting CCR.
But, the SHA fingerprint of the current EFTL staged zip binary is:
92d9b7ba0649b80b6717397b0441e01dd547a1cd60ad5c274b394ffb70ed3a31
This is different from the one described in CCR PR and TCK result page:
13712800a2dd23b025ccc9cd61604b3292cae6e1035e401ea540e60b18ab3ab0

**Comment by scottmarlow** [comment]
I see Jakarta EE 9 references in the TCK doc, so the SHA should probably change again if the TCK doc is corrected.

**Comment by markt-asf** [comment]
I found one more version update. If there is another WebSocket TCK build it would be good to include the PR below but it isn't essential.
https://github.com/eclipse-ee4j/jakartaee-tck/pull/922

**Comment by scottmarlow** [comment]
> I found one more version update. If there is another WebSocket TCK build it would be good to include the PR below but it isn't essential. [eclipse-ee4j/jakartaee-tck#922](https://github.com/eclipse-ee4j/jakartaee-tck/pull/922)

@markt-asf https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee10/staged/eftl/jakarta-websocket-tck-2.1.0.zip is now available with the eclipse-ee4j/jakartaee-tck#922](https://github.com/eclipse-ee4j/jakartaee-tck/pull/922 change.

Please note that with latest GlassFish we do see one failure https://ci.eclipse.org/jakartaee-tck/job/10/job/eftl-standalonetck-build-run-100/lastCompletedBuild/testReport/com.sun.ts.tests.websocket.ee.jakarta.websocket.remoteendpoint.async/WSClient/sendPingDelaysTimoutOnServerTest_from_standalone/.

**Comment by markt-asf** [comment]
I have no view on the CDI change since the implementation I work with doesn't use CDI. @jansupol is probably the best person to ask as any update to the TCK may require a re-run of TCK for certification purposes.

The WebSocket API milestones are published to the [staging repo](https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/websocket/). There are no plans to publish milestones to Maven Central as that tends to cause more problems than it solves.

**Comment by scottmarlow** [comment]
> I have no view on the CDI change since the implementation I work with doesn't use CDI. @jansupol is probably the best person to ask as any update to the TCK may require a re-run of TCK for certification purposes.
> 
> The WebSocket API milestones are published to the [staging repo](https://jakarta.oss.sonatype.org/content/groups/staging/jakarta/websocket/). There are no plans to publish milestones to Maven Central as that tends to cause more problems than it solves.

I looked closer at the Standalone Websocket TCK and that does include the CDI test that is now fixed, so we should build again.  Please expect one more websocket tck build update.  Hope that is okay.

**Comment by jansupol** [comment]
Standalone TCK does not contain CDI tests which are in the [src/com/sun/ts/tests/websocket/platform](https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/src/com/sun/ts/tests/websocket/platform folder). Only the full platform TCK contains the CDI tests. The standalone TCK should not be affected.

**Comment by scottmarlow** [comment]
> Standalone TCK does not contain CDI tests which are in the [src/com/sun/ts/tests/websocket/platform](https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/src/com/sun/ts/tests/websocket/platform) folder. Only the full platform TCK contains the CDI tests. The standalone TCK should not be affected.

Sorry, my bad.  https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee10/staged/eftl/jakarta-websocket-tck-2.1.0.zip was not updated (still has date of `2022-04-14`), so no new TCK was staged.

**Comment by kazumura** [comment]
@jansupol 
Could you update the CCR (https://github.com/jakartaee/websocket/issues/396) with the latest TCK ?

**Comment by kazumura** [comment]
> @kazumura @scottmarlow I have updated the CCR.

Thanks! @jansupol

**Comment by kazumura** [comment]
@markt-asf  
Could you send e-mail to [emo@eclipse.org](mailto:emo@eclipse.org) for release review and to [ee4j-pmc@eclipse.org](mailto:ee4j-pmc@eclipse.org) for release approval according to https://www.eclipse.org/projects/handbook/#release-review ?

**Comment by markt-asf** [comment]
Done.

File: websocket/2.1/_index.md [hunk 2]
```diff
@@ -1,46 +1,45 @@
...
+* [Jakarta WebSocket 2.1 TCK](https://download.eclipse.org/jakartaee/websocket/2.1/jakarta-websocket-tck-2.1.0.zip)  ([sig](https://download.eclipse.org/jakartaee/websocket/2.1/jakarta-websocket-tck-2.1.0.zip.sig),  [sha](https://download.eclipse.org/jakartaee/websocket/2.1/jakarta-websocket-tck-2.1.0.zip.sha256),  [pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
+* Maven coordinates
+    * [jakarta.websocket:jakarta.websocket-api:jar:2.1.0](https://search.maven.org/artifact/jakarta.websocket/jakarta.websocket-api/2.1.0/jar)
+    * [jakarta.websocket:jakarta.websocket-client-api:jar:2.1.0](https://search.maven.org/artifact/jakarta.websocket/jakarta.websocket-client-api/2.1.0/jar)
 
 # Compatible Implementations
 
-* Eclipse Tyrus. Version TBD
-* Apache Tomcat 10.1.x ?
+* TBD
```
Comment (ivargrimstad) [inline websocket/2.1/_index.md:23]: Add link to Tyrus here

**Comment by markt-asf** [comment]
Just a rebase

File: websocket/2.1/_index.md (hunk 2 above)
Comment (markt-asf) [reply websocket/2.1/_index.md:23]: Done

**Review by kazumura — APPROVED** [review]

**Comment by kazumura** [comment]
* On ballot completion, the specification committee mentor:
 - [x] adds this final checklist to the main PR.
 - [x] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
 \<include url of Issue here>

**Pull request merged**